### PR TITLE
Migrate to multipath-tcp.org generic netlink API.

### LINKED ISSUE #12: Migrate to multipath-tcp.org genl API
Replace the deprecated MPTCP generic netlink interface implementation in mptcpd with the one available in the multipath-tcp.org kernel.

### PR DESCRIPTION
**Pull request by ossama-othman**
Replace the deprecated MPTCP generic netlink interface implementation
in mptcpd with the one available in the multipath-tcp.org kernel.
Closes #12.

### CODE REVIEW

File: lib/plugin.c [hunk 1]
```diff
@@ -24,6 +24,19 @@
...
+ * @todo Remove this preprocessor symbol definition once support for
+ *       path managemnent strategy names are supported in the new
+ *       generic netlink API.
+ *
+ * @note @c GENL_NAMSIZ is used as the size since the path manager
+ *       name attribute in the deprecated MPTCP generic netlink API
+ *       contained a fixed length string of that size.
+ */
+#ifndef MPTCP_PM_NAME_LEN
+# include <linux/genetlink.h>  // For GENL_NAMSIZ
```
Comment (mjmartineau) [inline lib/plugin.c]: clang-format doesn't preserve this indentation style, fwiw

File: src/path_manager.c [hunk 2]
```diff
@@ -526,42 +698,136 @@ static void conn_closed_callback(struct l_genl_msg *msg, void *user_data)
...
+        mptcpd_plugin_subflow_priority(*token, &laddr, &raddr, backup, pm);
+}
+
+static void handle_mptcp_event(struct l_genl_msg *msg, void *user_data)
+{
+        int const cmd = l_genl_msg_get_command(msg);
+
+        assert(cmd != 0);
+
+        switch(cmd) {
```
Comment (mjmartineau) [inline src/path_manager.c]: Space missing after `switch`

File: include/linux/mptcp.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
+/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
```
Comment (mjmartineau) [inline include/linux/mptcp.h]: Thanks for getting this SPDX tag merged upstream!

File: lib/plugin.c (hunk 1 above)
Comment (ossama-othman) [reply lib/plugin.c]: I couldn't get clang-format to support this indentation style.  As it stands, the mptcpd `.clang-format` file brings code close the mptcpd coding style, but not exactly like it in some cases.

File: lib/plugin.c [hunk 4]
```diff
@@ -24,6 +24,19 @@
 
 #include <mptcpd/plugin.h>
 
+/**
+ * @todo Remove this preprocessor symbol definition once support for
+ *       path managemnent strategy names are supported in the new
```
[the remark this answers is not in the document]
Comment (ossama-othman) [reply lib/plugin.c]: Good catch.  I'll fix that.

File: src/path_manager.c (hunk 2 above)
Comment (ossama-othman) [reply src/path_manager.c]: Thanks.  I'll fix that.

File: tests/test-plugin.h [hunk 5]
```diff
@@ -41,50 +41,66 @@ struct plugin_call_count
...
-                && rhs->new_subflow >= 0
-                && rhs->subflow_closed >= 0
-                && rhs->connection_closed >= 0
-
-                && lhs->connection_closed <= rhs->new_connection
-                && lhs->subflow_closed <= rhs->new_subflow;
+        return     p->new_connection >= 0
+                && p->connection_established >= 0
+                && p->connection_closed >= 0
+                && p->new_address >= 0
```
Comment (ossama-othman) [inline tests/test-plugin.h]: I just noticed a missing `&& p->address_removed >= 0` condition here.  I'll add it.

File: src/path_manager.c [hunk 6]
```diff
@@ -342,99 +445,153 @@ static void new_subflow_callback(struct l_genl_msg *msg, void *user_data)
...
-                                  pm);
+        mptcpd_plugin_new_subflow(*token, &laddr, &raddr, backup, pm);
 }
 
-static void subflow_closed_callback(struct l_genl_msg *msg,
-                                    void *user_data)
+static void handle_subflow_closed(struct l_genl_msg *msg, void *user_data)
 {
+        /**
+         * @todo Refactor duplicate code in the handle_subflow_*()
```
[the remark this answers is not in the document]
Comment (ossama-othman) [reply src/path_manager.c]: The latest push contains the refactored code.  I do need to test it, however.

[the remark this answers is not in the document]
Comment (ossama-othman) [reply src/path_manager.c]: I should point out that the unit tests all pass, but they don't do a good job of testing the `src/path_manager.c` internals.

**Pull request merged**